File: src/ui/web/frontend/src/api/reports.js
/**
 * Reports API - Uses Gateway API
 * Handle content reports (moderation)
 */

import { get, post, put, del } from '@/api/client'
import { ENDPOINTS } from '@/api/config'
import { DEFAULTS } from '@/config/defaults'

// Report target types
export const ReportTargetType = {
  TEMPLATE: 'template',
  USER: 'user',
  REVIEW: 'review',
  MESSAGE: 'message'
}

// Report status
export const ReportStatus = {
  PENDING: 'pending',
  REVIEWING: 'reviewing',
  RESOLVED: 'resolved',
  DISMISSED: 'dismissed'
}

export const reportsAPI = {
  /**
   * Create a content report (requires auth)
   * @param {Object} data
   * @param {string} data.target_type - template, user, review, message
   * @param {string} data.target_id
   * @param {string} data.reason
   * @param {string} [data.details]
   * @returns {Promise<Object>}
   */
  async create(data) {
    try {
      const result = await post(ENDPOINTS.REPORTS.CREATE, data)
      return { ok: true, report: result.report }
    } catch (err) {
      return { ok: false, error: err.message }
    }
  },

  /**
   * List my submitted reports (requires auth)
   * @param {Object} [params]
   * @param {number} [params.page]
   * @param {number} [params.page_size]
   * @returns {Promise<Object>}
   */
  async listMyReports({ page = 1, page_size = DEFAULTS.PAGINATION.DEFAULT } = {}) {
    try {
      const result = await get(ENDPOINTS.REPORTS.MY_REPORTS, {
        params: { page, page_size }
      })
      return {
        ok: true,
        reports: result.reports || [],
        total: result.total || 0,
        page: result.page || page,
        pageSize: result.pageSize || pageSize
      }
    } catch (err) {
      return { ok: false, error: err.message, reports: [], total: 0 }
    }
  },

  // ============== Admin endpoints ==============

  /**
   * List all reports (admin only)
   * @param {Object} [params]
   * @param {string} [params.status]
   * @param {string} [params.target_type]
   * @param {number} [params.page]
   * @param {number} [params.page_size]
   * @returns {Promise<Object>}
   */
  async list({ status = null, target_type = null, page = 1, page_size = DEFAULTS.PAGINATION.DEFAULT } = {}) {
    try {
      const result = await get(ENDPOINTS.REPORTS.LIST, {
        params: { status, target_type, page, page_size }
      })
      return {
        ok: true,
        reports: result.reports || [],
        total: result.total || 0,
        page: result.page || page,
        pageSize: result.pageSize || pageSize
      }
    } catch (err) {
      return { ok: false, error: err.message, reports: [], total: 0 }
    }
  },

  /**
   * Get report by ID (admin only)
   * @param {string} reportId
   * @returns {Promise<Object>}
   */
  async get(reportId) {
    try {
      const result = await get(ENDPOINTS.REPORTS.GET(reportId))
      return { ok: true, report: result.report }
    } catch (err) {
      return { ok: false, error: err.message }
    }
  },

  /**
   * Get report statistics (admin only)
   * @returns {Promise<Object>}
   */
  async getStats() {
    try {
      const result = await get(ENDPOINTS.REPORTS.STATS)
      return { ok: true, ...result }
    } catch (err) {
      return { ok: false, error: err.message }
    }
  },

  /**
   * Update/resolve a report (admin only)
   * @param {string} reportId
   * @param {Object} data
   * @param {string} [data.status]
   * @param {string} [data.resolution_note]
   * @param {string} [data.action_taken]
   * @returns {Promise<Object>}
   */
  async update(reportId, data) {
    try {
      const result = await put(ENDPOINTS.REPORTS.UPDATE(reportId), data)
      return { ok: true, report: result.report }
    } catch (err) {
      return { ok: false, error: err.message }
    }
  },

  /**
   * Delete a report (admin only)
   * @param {string} reportId
   * @returns {Promise<Object>}
   */
  async delete(reportId) {
    try {
      await del(ENDPOINTS.REPORTS.DELETE(reportId))
      return { ok: true }
    } catch (err) {
      return { ok: false, error: err.message }
    }
  }
}

export default reportsAPI

File: src/ui/web/backend/api/reports.py
"""
Reports API

Endpoints for content reports (moderation).
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from typing import Optional

from api.auth import get_current_user
from gateway.providers.hub import get_provider_hub

router = APIRouter(prefix="/reports", tags=["reports"])


class CreateReportRequest(BaseModel):
    """Request body for creating a content report."""
    target_type: str  # template, user, review, message
    target_id: str
    reason: str
    details: Optional[str] = None


class UpdateReportRequest(BaseModel):
    """Request body for updating a report status."""
    status: Optional[str] = None  # pending, reviewing, resolved, dismissed
    resolution_note: Optional[str] = None
    action_taken: Optional[str] = None


# User endpoints

@router.post("/")
async def create_report(
    request: CreateReportRequest,
    user: dict = Depends(get_current_user),
):
    """Create a content report."""
    hub = get_provider_hub()

    try:
        from gateway.providers.data.models import ReportCreateDTO, ReportType

        report = await hub.data.reports.create_report(
            reporter_id=user["id"],
            data=ReportCreateDTO(
                target_type=ReportType(request.target_type),
                target_id=request.target_id,
                reason=request.reason,
                details=request.details,
            ),
        )

        return {"ok": True, "report": report.model_dump()}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/my-reports")
async def list_my_reports(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    user: dict = Depends(get_current_user),
):
    """List reports I've submitted."""
    hub = get_provider_hub()

    result = await hub.data.reports.get_user_reports(
        user_id=user["id"],
        page=page,
        page_size=page_size,
    )

    return {
        "ok": True,
        "reports": [r.model_dump() for r in result.items],
        "total": result.total,
        "page": result.page,
        "page_size": result.page_size,
    }


# Admin endpoints — MIGRATED to flyto-admin BFF on 2026-05-19.
#   GET    /reports            → flyto-admin GET    /admin/cloud/reports
#   GET    /reports/stats      → flyto-admin GET    /admin/cloud/reports/stats
#   GET    /reports/{id}       → flyto-admin GET    /admin/cloud/reports/{id}
#   PUT    /reports/{id}       → flyto-admin PATCH  /admin/cloud/reports/{id}
#   DELETE /reports/{id}       → flyto-admin DELETE /admin/cloud/reports/{id}
# Customer-facing endpoints above (POST /, GET /my-reports) stay here.
